Milldrill holes with helixes

chaotix- opened the pull request:
Implementation to milldrill holes with helixes as discussed in Issue #223.

Helix starts one step above Z0 for optimal work piece entry and makes a flat circle at the bottom.

This leads to two more passes in the for-loop: current_step = -1 at the beginning and current_step = stepcount at the end.

current_step = -1 unfortunately leads to a change of the types from "unsigned int" to "int" for stepcount and current_step. But I don't think this is a problem. Otherwise current_step could be transposed (-1 => 0, 0 => 1) and dropping the "+1" from the z height calculation, but I think this would make the code harder to understand.

And current_step = stepcount needs a min() call to prevent from cutting deeper than zwork. This feels a bit hacky but I thought it better than duplicating the output formatting for the arc and needs to be done only once even if/when helix slots get implemented (which would lead to a duplication of the more complicated slot formatting code otherwise).

I also removed the X and Y parameters from the arcs as I don't think they are necessary. Although it should not be a problem to add them again if some controllers do need them.

eyal0 commented:
The integration tests are going to fail.  You need to regenerate them.  The instructions are at the end of this: https://api.travis-ci.org/v3/job/441784448/log.txt


```
Run one of these:
./integration_tests.py --fix
./integration_tests.py --fix --add
```

Thanks!

chaotix- commented:
I'm a bit confused...

I used ./integration_tests.py before committing, but it threw 17 errors for the expected outputs of milldrill.ngc and drill.ngc in various tests. As drill.ngc should not be different with my patch I used ./integration_tests.py on an unpatched version and got the same 17 errors. So I thought the integration tests were not up-to-date.

Now Github is only reporting two errors with milldrill.ngc files and the changes in the output are just what I would expect with my patch. But running `./integration_tests.py --fix` on my computer tries to "fix" all 17 expected outputs which is probably nothing I should commit...

I guess there is something wrong at my end. But I don't know what or how to fix this. I never had problems with the generated files from pcb2gcode that I milled.

eyal0 commented:
The function for mill-drill covers a matrix of cases:

| Helix work            | Slot     | No slot |   
|-----------------------|----------|---------|
| Hole bigger than bit  | Not Done | Done    |
| Hole smaller than bit | Not Done | N/A     |


Ideally we would fix them all.  It wouldn't really be THAT hard to fix the slot case, too.  Are you willing?  If you did, you wouldn't need the special case for `!slot` to set the step to -1.

For slots where the bit is to big, it would just be a zig-zag working the way down and the final zag would be at constant Z, just like you did.  Really easy to do, you just combine the XY command and the following Z command.

For oval slots, each step is four commands: half-circle, across, half-circle, and then across again.  The path isn't really an oval, it's the convex hull of two circles.  The total path length is just the circumference of the circle plus the distance between the start and end centers doubled.  Each part of the oval gets a portion of the Z travel depending on how long it is.

A little more work but since you've already started, maybe you're interested?  Let me know.  Either way, the integration tests need fixing.

eyal0 commented:
Okay, if you're having trouble with the integration tests, it could be due to the version of gerbv or boost that you are using.  Unfortunately, every tiny change in gerbv or boost causes changes in the output.  So only boost 1.66 and gerbv 2.6A generate the integration outputs correctly.  Kind of sucks but it's the best that I could do and it's been useful in catching errors so far.

if you're using linux, you should be able to patch in the broken log like this:

`curl https://api.travis-ci.org/v3/job/441784448/log.txt | patch -p1`

chaotix- commented:
OK. I've added back the X and Y parameters as you prefer them. Once the integration test fails again I will patch the expected results it with the updated output.

(I'm using Debian stretch with boost 1.62 and gerbv 2.6.1. But it's not just a rounding error here and there, there are many, many holes missing in my output. Never had a problem with that with my PCBs. Weird.)

eyal0 commented:
If you tell me which holes are missing, maybe I could figure it out.  Could it be that your version of gerbv doesn't support slots and all the slots are missing?  I have a few tests around that stuff.

chaotix- commented:
Here is the output of `./integration_tests.py 2> /tmp/test.log`

[test.log](https://github.com/pcb2gcode/pcb2gcode/files/2480153/test.log)

eyal0 commented:
> Here is the output of `./integration_tests.py 2> /tmp/test.log` (from your master branch without my patches)
> 
> [test.log](https://github.com/pcb2gcode/pcb2gcode/files/2480153/test.log)

I'm pretty sure that all those errors are due to a lack of slot support in your version of gerbv.  Slot support is in 2.6A and 2.6A comes *after* 2.6.1, surprisingly.  If you open the gcode in https://ncviewer.com/ and compare the files that change both before and after, you'll see the 3 slots that are drilled into the multivibrator example.

If you don't want to install gerbv form sources and keep using the gerbv that you're using, just run that curl command from above on the latest failure, which will be here: https://travis-ci.org/pcb2gcode/pcb2gcode/builds/441815366?utm_source=github_status&utm_medium=notification  You need to wait for the test to fail and then download the raw log and use the curl command like I showed above.

chaotix- commented:
Ah. I was using a self-written Python preprocessor for the drill files before switching from GRML/bCNC to LinuxCNC. The preprocessor rounded the drill bits to full 0,1mm sizes and approximated slots with a bunch of simple holes. I wrote this before you implemented those features in pcb2gcode.

When I switched to LinuxCNC I had to update my workflow as the autolevelling is different. In the course of this I updated pcb2gcode and saw that it now supports proper slots and rounding of drill bit sizes. But I have never milled a board with those features so I have never used the slot support in pcb2gcode which explains why I never had problems before with my outdated gerbv version.

eyal0 commented on testing/gerbv_example/slots-milldrill/expected/milldrill.ngc:
```diff
@@ -577,558 +577,303 @@ G1 X3.95183 Y-2.39173
-G2  X4.46063 Y-2.85000 I-0.00906 J0.00000
-G1 Z-1.60000
-G2  X4.46063 Y-2.85000 I-0.00906 J0.00000
+G1 Z0.02963
```
Should this be a G0?  G0 is for fast moves that are not doing any drilling.  What do you think?

chaotix- replied:
I opted for the G1 because it's getting very close to the work piece and it's not autolevelled. So it could hit the surface if the PCB is uneven which I don't think is a good idea with G0.

eyal0 replied:
Good call.

eyal0 commented on testing/gerbv_example/slots-milldrill/expected/milldrill.ngc:
```diff
@@ -577,558 +577,303 @@ G1 X3.95183 Y-2.39173
+G2 X4.46063 Y-2.85000 Z-1.54074 I-0.00906 J0.00000
+G2 X4.46063 Y-2.85000 Z-1.57037 I-0.00906 J0.00000
+G2 X4.46063 Y-2.85000 Z-1.60000 I-0.00906 J0.00000
+G2 X4.46063 Y-2.85000 Z-1.60000 I-0.00906 J0.00000
```
We don't need the Z on the second time around.  The downside is that someone looking at it might think that the line was repeated on accident and remove one of them.  But making a special case for this makes the code longer and more ugly so I think that it's okay to leave it like this.

chaotix- replied:
Yeah. I thought so too and was too lazy for the special case. :-)

But I implemented something and it doesn't look that bad:
```
          if (!slot) {
            // Just drill a full-circle.
            of << arc_gcode
               << " X" << start_targetx * cfactor
               << " Y" << start_targety * cfactor;
            // Drop superflous Z from the bottom hole to indicate that this line is not there by accident
            if(stepcount != current_step) {
               of << " Z" << z * cfactor;
            }
            of << " I" << (start_x-start_targetx) * cfactor
               << " J" << (start_y-start_targety) * cfactor << "\n";
          } else {
```

chaotix- replied:
OK. Committed. Will update the tests once they have failed and I can get the patch.

I will see about the slots, but I'm not making any promises as I don't know if I will have the time.

And I need to fix the slot problem on my side before that. I hope it's just the outdated gerbv version. I don't want to update libboost on my computer as it has too many dependencies...

eyal0 replied:
I think that it's just gerbv that needs building because only the slots looked wrong.  You can see how gerbv is installed here:

https://github.com/pcb2gcode/pcb2gcode/blob/master/.travis.yml#L79

If you're unable to also support the slots, that's okay, too.  I might get around to it.

eyal0 commented:
I just realized that I totally didn't think about feedrate. Pcb2gcode has a vertical drill feedrate and also as horizontal one. This is going in three axes so I dunno what the right speed is. I guess the minimum of the two? We want the feedrate set so that the vertical change over time is not more than the maximum vertical speed and the horizontal cutting motion also does not exceed the maximum.

chaotix- commented:
Oh, yeah. Forgot about that as well.

I just checked the FreeCAD source code and they use there horizontal feedrate for helix cuts. Because I didn't believe this to be right I also made a slim hole in Fusion 360, selected a cutter nearly the size of the hole (+0,1mm) with a big pitch and checked there. They also use the horizontal (or cutting as they call it) feedrate although the path is more or less vertical. So maybe it is OK to do so simply?

We could of course constrain the feedrate to be within the horizontal and vertical feedrate, but as both FreeCAD and Fusion 360 don't do that, I'm not sure this is necessary. Considering that the holes probably are always less then 2mm deep it might be limited by the acceleration settings of the machine anyway...

chaotix- commented:
Fortunately I had some time today. I added simple feedrates (vertical feedrate for the initial drop before starting the real cutting as it might touch the surface and horizontal rate for the helix/zigzag) and slot support.

When updating the integration tests (BTW, you were right, my problem was just the outdated gerbv)  I noticed that all/a lot of millproject files were ... unusual. They use inches but have zdrill = -1.6 which looks weird when looking at the resulting files with CAMotics.

eyal0 commented:
Oh, I see.  Yes, the integration tests looks pretty worthless when the zigzag is set with so deep a cutinfeed.  I should fix those first to be a more interesting value, that way looking that the [diffs](https://github.com/pcb2gcode/pcb2gcode/pull/224/files) will be more interesting!  Thanks for noticing that.

chaotix- commented:
What do you think about this? It's nice and simple.

eyal0 commented:
I updated the integrations tests for you and merged master.  Had a complex time trying to get git to do what I want, too!  I'll take a look.

eyal0 commented on drill.cpp:
```diff
@@ -383,32 +383,48 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
+                   << '\n';
             }
         } else {
             of << "G1 Z" << cutter->zwork * cfactor << '\n';
```
Now that we're messing around with the Feedrate so often, maybe we need to be explicit here.  Because we can't count on the most recent feedrate to be for a plunge.  Am I right?

chaotix- replied:
Oh. My mistake. I thought it was just always the same, but I removed one. I will fix this.

eyal0 commented on drill.cpp:
```diff
@@ -383,32 +383,46 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
+
+            // Is there enough room for material evacuation?
+            if (distance > 0.3 * cutdiameter) {
+                of  << "F" << cutter->feed * cfactor << '\n';
```
Two feedrate lines in a row is not as pretty in the output.  Why not just combine this output and the one above it into if-else?

chaotix- replied:
I don't understand. This is not on the same line.

This comes out as
```
G0 X3.81500 Y-2.66535
G1 Z0.02100 F50.00000
F100.00000
G1 X3.81500 Y-2.58665 Z0.01050
```

So it does the initial plunge with F50, then changes the feedrate to F100 for the rest of the moves if the hole/slot is big enough for the horizontal feedrate).

eyal0 replied:
Ah, okay, now I follow!  Right, that's makes sense.  I forgot that the F50 actually affects the text on the same line, too, so it is actually setting the feedrate for that G1 and then setting it again.  Cool.

eyal0 commented on drill.cpp:
```diff
@@ -383,32 +383,46 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
+                   << '\n';
             }
         } else {
             of << "G1 Z" << cutter->zwork * cfactor << '\n';
```
But what about this line?  If we milldrill a slot and then milldrill all the way through, this G1 will use the most recent feedrate, right?  So it'll be at the horizontal rate instead of the vertical one.  Right?

Just need to add an "F" command there.  It's always vertical.

chaotix- replied:
You are right. There must be an explicit feedrate there. I will prepare a pull request.

eyal0 commented:
Great work.  Almost done, I think!
